src/client.ts

```diff
@@ -16,109 +43,214 @@ export type ResourceIteratee<T> = (each: T) => Promise<void> | void;
  * resources.
  */
 export class APIClient {
-  constructor(readonly config: IntegrationConfig) {}
+  constructor(
+    readonly config: IntegrationConfig,
+    readonly logger: IntegrationLogger,
+  ) {}
+
+  private generateAuthToken(
+    dateString: string,
+    reqId: string,
+    uri: string,
+  ): string {
+    const { clientId, clientSecret, appKey } = this.config;
+
+    const dataToSign = `${dateString}:${reqId}:${uri}:${appKey}`;
+    const hmac = createHmac('sha1', Buffer.from(clientSecret, 'base64'));
+    hmac.update(dataToSign);
+    const signed = hmac.digest('base64');
+
+    return `MC ${clientId}:${signed}`;
+  }
+
+  private generateHeaders(uri: string) {
+    const { appId } = this.config;
+    const now = new Date();
+    const nowUTCString = now.toUTCString();
+    const nowReallyUTCString =
+      nowUTCString.slice(0, nowUTCString.length - 3) + 'UTC';
+    const reqId = uuid();
+    return {
+      'x-mc-date': nowReallyUTCString,
+      'x-mc-req-id': reqId,
+      'x-mc-app-id': appId,
+      Authorization: this.generateAuthToken(nowReallyUTCString, reqId, uri),
+    };
+  }
+
+  // route-agnostic response validation
+  private verifyResponse(response: ApiResponse<any>, endpoint: string) {
+    if (response.meta.status !== 200) {
+      const toThrow = new IntegrationProviderAPIError({
+        message: `encountered non-200 status code in response body despite getting a 200 in request lib: ${JSON.stringify(
+          response,
+        )}`,
+        endpoint,
+        status: response.meta.status,
+        statusText: statusTextMap[response.meta.status] || 'unknown',
+      });
+      throw toThrow;
+    }
+    if (response.fail && response.fail.length) {
```

Comment:
This is the case that is going to get retried, right? Might be nice to add a small comment here as a reminder